enhancements/20240805-encrypted-block-volume.md

```diff
@@ -0,0 +1,335 @@
+# Support Block Volume Encryption/Decryption
+
+This enhancement focuses on the encryption and decryption of block volumes, particularly involving boot volumes and data volumes. 
+
+
+## Summary
+
+This enhancement introduces encryption and decryption capabilities for block volumes, specifically targeting boot volumes (virtual machine images) and data volumes. This functionality will enhance security by ensuring that data stored in these volumes is encrypted and can be decrypted when attached. 
+
+## Related Issues
+
+https://github.com/harvester/harvester/issues/3129
+
+## Motivation
+
+
+### Goals
+
+- Implement encryption for boot volumes and data volumes (support hot plug). 
+- Provide a clear and detailed setup process for users to enable encryption/decryption. 
+
+### Non-goals
+
+- This enhancement will not cover GUI-related changes. That particularly means we won't provide more convenient way to enable encryption/decryption with GUI in this enhancement. This will be addressed in next enhancement.
+
+
+## Proposal
+
+Since Longhorn uses clone mechanism to encrypt/decrypt Boot volume, so Harvester will follow that. This means that there would be two virtual machine image after encryption. One is encrypted, and the other is source virtual machine image. 
+
+![](./20240805-encrypted-block-volume/lh-flow.png)
+
+Data volumes are different from Boot volume. Because Longhorn supports block encryption/decryption, so Harvester could encrypt/decrypt data volume directly. 
+
+Regardless of volumes used, Longhorn will decrypt volume when it's attached to virtual machine. So, we'll be able to read the boot volumes and data volumes in the virtual machine. 
+
+### User Stories
+
+#### Story 1 - Use encrypted boot volume to start virtual machine
+
+Alice wants to encrypt the current boot volume for security, so Alice follows up the guideline:
+
+1. Prepare a virtual machine image, which you could upload or use URL to download.
+2. Create a secret and template storage class for encryption/decryption.
+3. Create a new virtual machine image with encryption enabled.
+
+After that, Alice gets an encrypted virtual machine image as boot volume. Then, Alice can start the virtual machine with selecting encrypted boot volume in GUI. For security, Alice decides to delete that original virtual machine image. For this moment, Alice only has this encrypted virtual machine image and Alice can reuse this boot volume.
+
+Over time, Alice wants to decrypt the encrypted virtual machine image. Alice follows up the guideline:
```

Comment:
For security consideration, we may add option when encrypt an image:
With pass/key to decrypt
Always allow to decrypt

Then if some user (maybe not the owner) plans to decrypt certain image, he needs to provide proof.

Reply:
I think it's a complicated topic.

Since LH only provides k8s secret as key management, it's not enough for encryption/decryption at this moment. For more securities, we might need more different key managements.

Although we could force user to type the key in GUI each time when they'd like to encrypt/decrypt, but there are still issues by manually creating yaml. 

Besides, if we'd like to control the encrypt and decrypt permission, we probably need sub resource to define permission.

Reply:
No worries. We can claim cretain features are not covered in this HEP, and add related information and the possible workaround in the coming document. User will know we have considerations on such.

Reply:
Yeah, it's good point. Regardless of ways we choose, I think we should end up providing a full guideline about how to secure the volume. I'll keep in mind while I'm doing this feature.

Reply:
BTW, I'll add this part into enhancement as well, thanks.